pkg/api/entries.go

```diff
@@ -411,46 +394,51 @@ func SearchLogQueryHandler(params entries.SearchLogQueryParams) middleware.Respo
 	if len(params.Entry.LogIndexes) > 0 {
 		g, _ := errgroup.WithContext(httpReqCtx)
 
-		leafResults := make([]*trillian.GetEntryAndProofResponse, len(params.Entry.LogIndexes))
-		for i, logIndex := range params.Entry.LogIndexes {
-			i, logIndex := i, logIndex // https://golang.org/doc/faq#closures_and_goroutines
+		for _, logIndex := range params.Entry.LogIndexes {
+			logIndex := logIndex // https://golang.org/doc/faq#closures_and_goroutines
 			g.Go(func() error {
-				tid, resolvedIndex := api.logRanges.ResolveVirtualIndex(int(swag.Int64Value(logIndex)))
-				trillianClient := NewTrillianClientFromTreeID(httpReqCtx, tid)
-				resp := trillianClient.getLeafAndProofByIndex(resolvedIndex)
-				switch resp.status {
-				case codes.OK, codes.NotFound:
-				default:
-					return resp.err
-				}
-				leafResult := resp.getLeafAndProofResult
-				if leafResult != nil && leafResult.Leaf != nil {
-					leafResults[i] = leafResult
+				logEntry, err := retrieveLogEntryByIndex(httpReqCtx, int(swag.Int64Value(logIndex)))
+				if err != nil {
+					return err
 				}
+				resultPayload = append(resultPayload, logEntry)
 				return nil
 			})
 		}
 
 		if err := g.Wait(); err != nil {
 			return handleRekorAPIError(params, http.StatusInternalServerError, fmt.Errorf("grpc error: %w", err), trillianUnexpectedResult)
 		}
-
-		for _, result := range leafResults {
-			if result != nil {
-				logEntry, err := logEntryFromLeaf(httpReqCtx, api.signer, tc, result.Leaf, result.SignedLogRoot, result.Proof, api.logRanges.ActiveTreeID(), api.logRanges)
-				if err != nil {
-					return handleRekorAPIError(params, http.StatusInternalServerError, err, trillianUnexpectedResult)
-				}
-				resultPayload = append(resultPayload, logEntry)
-			}
-		}
 	}
 
 	return entries.NewSearchLogQueryOK().WithPayload(resultPayload)
 }
 
 var ErrNotFound = errors.New("grpc returned 0 leaves with success code")
 
+func retrieveLogEntryByIndex(ctx context.Context, logIndex int) (models.LogEntry, error) {
+	tid, resolvedIndex := api.logRanges.ResolveVirtualIndex(logIndex)
+	tc := NewTrillianClientFromTreeID(ctx, tid)
+	log.ContextLogger(ctx).Debugf("Retrieving resolved index %v from TreeID %v", resolvedIndex, tid)
+
+	resp := tc.getLeafAndProofByIndex(resolvedIndex)
+	switch resp.status {
+	case codes.OK:
+	case codes.NotFound, codes.OutOfRange, codes.InvalidArgument:
+		return models.LogEntry{}, ErrNotFound
+	default:
+		return models.LogEntry{}, fmt.Errorf("%s: %w", trillianCommunicationError, resp.err)
```

Comment:
WDYT of:


```go
return models.LogEntry{}, fmt.Errorf("grpc err: %w: %s", resp.err, trillianCommunicationError)
```

Since this is a helper function I'd prefer to avoid using `handleRekorAPIError`. I think this should contain all the same info in the error message and return the same status code.

Reply:
Ah! Good point, yeah, actually then just on line 402, wrap the `err` with a `handleRekorAPIError(params, http.StatusInternalServerError,...)`?

Reply:
updated 😄